cert/aziot-certd/src/lib.rs

```diff
@@ -195,654 +258,349 @@ impl UpdateConfig for Api {
+                        }
 
-                CertIssuanceMethod::SelfSigned => {
-                    // Since the client did not give us their private key handle, we assume that the key is named the same as the cert.
-                    //
-                    // TODO: Is there a way to not have to assume this?
-                    let key_pair_handle = api
-                        .key_client
-                        .load_key_pair(id)
-                        .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
-
-                    // Recurse with explicit issuer.
-                    let x509 = create_cert(api, id, csr, Some((id, &key_pair_handle))).await?;
-                    Ok(x509)
+                        let handle = api.key_client.create_key_pair_if_not_exists(
+                                &auth_x509.identity.pk,
+                                Some("ec-p256:rsa-4096:*")
+                            )
+                            .and_then(|handle| Ok(CString::new(handle.0)?))?;
+
+                        let id_pubkey = api.key_engine.load_public_key(&handle)?;
+                        let id_privkey = api.key_engine.load_private_key(&handle)?;
+
+                        let subject_name = build_name(
+                                cert_options.subject.as_ref()
+                                    .unwrap_or(&CertSubject::CommonName("est-id".to_owned()))
+                            )?;
+
+                        let mut builder = X509Req::builder()?;
+                        builder.set_version(0)?;
+                        builder.set_subject_name(&subject_name)?;
+
+                        let mut exts = Stack::new()?;
+                        exts.push(
+                                extension::ExtendedKeyUsage::new()
+                                    .client_auth()
+                                    .build()?
+                            )?;
+
+                        builder.add_extensions(&exts)?;
+                        builder.set_pubkey(&id_pubkey)?;
+                        builder.sign(&id_privkey, MessageDigest::sha256())?;
+
+                        let id_init = est::create_cert(
+                                base64::encode(builder.build().to_der()?).into_bytes(),
+                                url,
+                                auth.basic.as_ref(),
+                                Some((&bcert, &bpk)),
+                                &trusted_certs,
+                                api.proxy_uri.clone()
+                            )
+                            .await?;
+                        write_cert(
+                                &api.homedir_path,
+                                &api.preloaded_certs,
+                                &auth_x509.identity.cert,
+                                &id_init
+                            )?;
+
+                        create_cert_inner(api, id, (req, pubkey), issuer)
```

Comment:
Restore the original comment that explains why this recursive call is happening.